autoscale-runner.sls

```diff
@@ -0,0 +1,66 @@
+# This state installs a gitlab runner which is designed to act as a bastion node
+# for an autoscaling setup using docker-machine.
+# See [https://github.com/chr4/salt-docker] for some useful substates to prepare
+# the required docker base setup. In particular see states
+#   - docker
+#   - docker.login
+#   - docker.machine
+# In addition to installing the runner binary itself it supports the installation
+# of a set of pre-registered runner configurations by populating the config.toml.
+
+# Use the official Gitlab apt repository
+gitlab-runner-repo:
+  pkgrepo.managed:
+    - name: deb https://packages.gitlab.com/runner/gitlab-runner/{{ grains['os']|lower }}/ {{ grains['oscodename'] }} main
+    - file: /etc/apt/sources.list.d/runner_gitlab-runner.list
+    - key_url: https://packages.gitlab.com/gpg.key
+
+# Ensure that the gitlab-runner service terminates gracefully when stopped
+# by systemd. See https://docs.gitlab.com/runner/configuration/init.html#overriding-systemd
+# The long (2h) timeout ensures that started CI jobs are allowed to run to completion.
+/etc/systemd/system/gitlab-runner.service.d/kill.conf:
+  file.managed:
+    - user: root
+    - group: root
+    - mode: 644
+    - makedirs: True
+    - dir_mode: 755
+    - contents: |
+        [Service]
+        TimeoutStopSec=7200
+        KillSignal=SIGQUIT
+
+# Install the runner
+gitlab-runner:
```

Comment:
I think this can be done by using an `include` with the regular `runner.sls` resp. `runner/init.sls` state. The `require` parts (if they are really required to be present BEFORE installation of the package, which I doubt) can be also defined by the required states using `require_in`.

Reply:
I think we need some further discussion reg. the best dependency ordering/constraints.

The key thing that must be guaranteed is that
- the docker-machine certificate and key files (the last four lines of the current require-block) must be in place
- BEFORE the gitlab-runner service executes with a config that specifies a docker-machine executor (*).

The problem here appears to be that when the gitlab runner service launches with a docker-machien executor config it immediately requests the creation of multiple docker workers. If the underlying machine executor is at this point uninitialised (i.e. the key and cert files are not in place) the initialisation is also started multiple times in parallel. And that latter process is neither transactional, nor idempotent, so multiple parallel initialisation processes clobber each other and result in a non-deterministically inconsistent state, that requires wiping everything and starting from scratch.

(*) a config that contains this somewhere:

```
...
[[runners]]
  ...
  executor = "docker+machine"
  ...
...
```

The present setup was the result after quite some trial and error to ensure that the above two points always happen in the right order. And to the best of my knowledge it is not possible in salt to install a package and deploy its config _but prevent the associated service from starting_ (killing the service after the package installation started it is possible, but that is not good enough).

It is very well possible that fewer requirements are possible and I am happy to refactor/reorganise the state(s) along the lines suggested here and in other comments, but I don't think that we can get fully rid of a dependency link that connects this repository to https://github.com/chr4/salt-docker. In addition, I choose to place `require` statements here, rather than `require_in` statements in the docker repo, since

- `require_in` repeatedly had bugs in the past and appeared to be generally more unstable than `require` (i.e. simply ignoring the requested order)
- the docker-machine substate is the structurally lower level in the stack and in principle usable without ever having the gitlab runner around
- the need for the dependency is determined by the content of the gitlab-runner config. The latter logically belongs to this state and checking a pillar key for a gitlab runner config from a docker sub-state to determine if an upwards `require_in` dependency is needed just felt wrong to me. Note that we likely have to introduce such a non-trivial check for specific pillar content if we want to merge the proposed state with the existing `runner.sls`.